### ISSUE TITLE
How to choose the best model?

### ISSUE THREAD

**Issue by ppbangKGT** [gigo-team/bev_lane_det#5]
Thanks for your open-source codes. When I train the model, I find that it seems that you don't use tensorboard to record the loss or F1-score. And the total performance of one epoch isn't recorded either. So I wonder how to choose the best model I have got. Should I run the val_openlane.py on each of the models I trained? Or maybe I miss something important.
Thanks for your reply.

**Comment by rhwang1314** [comment]
Sorry, we only provided a simple version of the code. You can add tensorboard to your code according to your needs. In fact, we usually choose the latest model for testing.

**Comment by ppbangKGT** [comment]
> Sorry, we only provided a simple version of the code. You can add tensorboard to your code according to your needs. In fact, we usually choose the latest model for testing.

I have trained 30 epochs. I remember the result in your paper is gotten in 10 epochs on OpenLane dataset.
So is that mean you use the 10th trained model for testing? And it gets the best performance?
Thanks for your reply.

**Comment by rhwang1314** [comment]
We found that the effects of the model  of 10th epoch and models  of continue training were not significantly different.

**Comment by ppbangKGT** [comment]
And I find it seems that your codes can't resume training though there is a funtion called "resume_training".

**Comment by FeiYull** [comment]
@ppbangKGT  it seems so

**Comment by vanAken** [comment]
There are some small changes to do to use it:
in 
bev_lane_det/models/util/save_model.py

def save_model_dp
 make sure 
 
 torch.save({
        "**model_state**": net.module.state_dict(),
        "**optimizer_state**": optimizer.state_dict() if optimizer else None,
    }, model_path)
has the same names in 
bev_lane_det/models/util/load_model.py

def load_model(model, model_state_file):
    pretrained_dict = torch.load(model_state_file, map_location='cpu')
    if 'model_state' in pretrained_dict:
        pretrained_dict1 = pretrained_dict['**model_state**']
    elif 'state_dict' in pretrained_dict:
        pretrained_dict1 = pretrained_dict['**state_dict**']

**Comment by ppbangKGT** [comment]
> There are some small changes to do to use it: in bev_lane_det/models/util/save_model.py
> 
> def save_model_dp make sure
> 
> torch.save({ "**model_state**": net.module.state_dict(), "**optimizer_state**": optimizer.state_dict() if optimizer else None, }, model_path) has the same names in bev_lane_det/models/util/load_model.py
> 
> def load_model(model, model_state_file): pretrained_dict = torch.load(model_state_file, map_location='cpu') if 'model_state' in pretrained_dict: pretrained_dict1 = pretrained_dict['**model_state**'] elif 'state_dict' in pretrained_dict: pretrained_dict1 = pretrained_dict['**state_dict**']

Yes, there can be different methods to resume training. What I do is :
1. still use resume_training and load_checkpoint functions.
2. make sure the 'model_state' and 'optimizer_state' in load_checkpoint function are the same as those in save_model_dp function.
3. scheduler.step() seems to change to the epoch which I have trained.
The load_model function is only used in validation code originally. So I don't change it.

**Comment by idrivingman** [comment]
>hi,u set scheduler.step(x) where x is an int?like 9?

**Comment by EnternalTwinkle** [comment]
@ppbangKGT Can you share the source code? Thank you very much，1017094591@qq.com

**Comment by dagongji10** [comment]
@ppbangKGT hello, would you please share me the source code? 405612048@qq.com

**Comment by chenyuanpengcyp** [comment]
@ppbangKGT hello, would you please share me the source code? 1653658300@qq.com

**Comment by hitbuyi** [comment]
@ppbangKGT hello, would you please share me the source code? [hitbuyi@163.com](mailto:hitbuyi@163.com)

**Comment by shupinghu** [comment]
Hello, I am a new researcher of 3D-BEV-LaneDet, could you please share the source code with me?
My Email [hushp3@qq.com](mailto:hushp3@qq.com)
